File: libs/power-log-parser/src/lib/parsers/new-game-parser.ts
import { ActionParser } from '../action-parser';
import { GameEventProvider } from '../game-event';
import { Game, Node } from '../models';
import { Logger } from '../logger';
import { Utility } from '../utility';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class NewGameParser implements ActionParser {
	readonly ParserName = 'NewGameParser';

	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return stateType === StateType.GameState && node.Type === Game && !this.ParserState.ReconnectionOngoing;
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		Logger.Log('will emit NEW_GAME event', node.CreationLogLine);
		return [
			GameEventProvider.Create(
				(node.Object as Game).TimeStamp,
				'NEW_GAME',
				() => ({
					Type: 'NEW_GAME',
					Value: {
						Spectating: this.StateFacade.Spectating,
						Timestamp: Utility.GetUtcTimestamp((node.Object as Game).TimeStamp),
					},
				}),
				false,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/NewGameParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using HearthstoneReplays.Events;
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;

namespace HearthstoneReplays.Events.Parsers
{
    public class NewGameParser : ActionParser
    {
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public NewGameParser(ParserState ParserState, StateFacade StateFacade)
        {
            this.ParserState = ParserState;
            this.StateFacade = StateFacade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.GameState
                && node.Type == typeof(Game)
                && !ParserState.ReconnectionOngoing;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            Logger.Log("will emit NEW_GAME event", node.CreationLogLine);
            return new List<GameEventProvider> { GameEventProvider.Create(
                (node.Object as Game).TimeStamp,
                "NEW_GAME",
                () => new GameEvent
                {
                    Type = "NEW_GAME",
                    Value = new
                    {
                        Spectating = StateFacade.Spectating,
                        Timestamp = Utility.GetUtcTimestamp((node.Object as Game).TimeStamp)
                    },
                },
                false,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
